src/client/dfuse/ops/read.c

```diff
@@ -12,159 +12,160 @@
 static void
 dfuse_cb_read_complete(struct dfuse_event *ev)
 {
-	if (ev->de_ev.ev_error == 0) {
-		size_t len = ev->de_iov.iov_buf_len;
-
-		if (ev->de_len == 0)
-			DFUSE_TRA_DEBUG(ev, "Truncated read, (EOF)");
-		else if (ev->de_len != len)
-			DFUSE_TRA_DEBUG(ev,
-					"Truncated read, requested %#zx returned %#zx",
-					len, ev->de_len);
+	if (ev->de_ev.ev_error != 0) {
+		DFUSE_REPLY_ERR_RAW(ev, ev->de_req, ev->de_ev.ev_error);
+		D_GOTO(free, 0);
+	}
 
+	if (ev->de_len == 0) {
+		DFUSE_TRA_DEBUG(ev, "Truncated read, (EOF)");
 		DFUSE_REPLY_BUF(ev, ev->de_req, ev->de_iov.iov_buf, ev->de_len);
-	} else {
-		DFUSE_REPLY_ERR_RAW(ev, ev->de_req, ev->de_ev.ev_error);
+		D_GOTO(free, 0);
+	}
+
+	if (ev->de_len < ev->de_req_len)
+		DFUSE_TRA_DEBUG(ev, "Truncated read, requested %#zx returned %#zx", ev->de_req_len,
+				ev->de_len);
+	else if (ev->de_len < ev->de_iov.iov_buf_len)
+		DFUSE_TRA_DEBUG(ev, "Truncated readhead, requested %#zx returned %#zx",
+				ev->de_iov.iov_buf_len, ev->de_len);
+
+	if (ev->de_len > ev->de_req_len) {
+		struct dfuse_obj_hdl *oh;
+		int                   rc;
+
+		oh = ev->de_oh;
+		rc = pthread_mutex_trylock(&oh->doh_ie->ie_dfs->dfs_read_mutex);
```

Comment:
Does this mean it's possible for the kernel to request a read for data that hasn't been cached yet, and we end up calling `dfs_read` for that data again?

Reply:
No, it doesn't mean anything that complicated.  The trylock is to avoid all threads trying to write in the kernel buffers concurrently and blocking each other.  Given that this now happens in the callback it'll probably be called in the event handler so all by one thread so maybe that's less of an issue how.  The intent was to avoid blocking on locks when dfuse had replies in-hand which the kernel had requested.

Reply:
What isn't clear is that if `pthread_mutex_trylock` fails because the lock is already held, do we retry caching later? At the end of `dfuse_cb_read_complete` we free the buffer, so it seems if the lock was already held, we just skip caching the readahead data. What am I missing here?

Reply:
No, we just drop it.  It's important that we populate the kernel buffer before replying to the request or the kernel will likely immediately make a request for the next data just as we're populating the buffer, and that leads to the read being performed twice.

Reply:
> that leads to the read being performed twice

Ah, that's what I was curious about. Thanks. Maybe a discussion for a future PR, but in the extreme case, it seems this could happen:
```
file size is 1280K (1M + 256K)
kernel requests 256K
we readahead 256K + 1M
lock was held, so we don't cache and just return 256K
kernel requests next 256K
we readahead 256K + 1M
lock was held, so we don't cache and just return 256K
...
total read = (256K + 1M) * 5
file size  = (256K + 1M) * 1
overhead of 5x!
```

Reply:
Yes, that could possibly happen as well.